watchtower/wtclient/client.go

```diff
@@ -876,6 +877,125 @@ func (c *TowerClient) handleChannelCloses(chanSub subscribe.Subscription) {
 	}
 }
 
+// deleteSessionFromTower dials the tower that we created the session with and
+// attempts to send the tower the DeleteSession message.
+func (c *TowerClient) deleteSessionFromTower(sess *wtdb.ClientSession) error {
+	// First, we check if we have already loaded this tower in our
+	// candidate towers iterator.
+	tower, err := c.candidateTowers.GetTower(sess.TowerID)
+	if errors.Is(err, ErrTowerNotInIterator) {
+		// If not, then we attempt to load it from the DB.
+		dbTower, err := c.cfg.DB.LoadTowerByID(sess.TowerID)
+		if err != nil {
+			return err
+		}
+
+		tower, err = NewTowerFromDBTower(dbTower)
+		if err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+
+	session, err := NewClientSessionFromDBSession(
+		sess, tower, c.cfg.SecretKeyRing,
+	)
+	if err != nil {
+		return err
+	}
+
+	localInit := wtwire.NewInitMessage(
+		lnwire.NewRawFeatureVector(wtwire.AltruistSessionsRequired),
+		c.cfg.ChainHash,
+	)
+
+	var (
+		conn      wtserver.Peer
+		towerAddr = tower.Addresses.Peek()
+	)
+	// Attempt to dial the tower with its available addresses.
+	for {
+		conn, err = c.dial(
+			session.SessionKeyECDH, &lnwire.NetAddress{
+				IdentityKey: tower.IdentityKey,
+				Address:     towerAddr,
+			},
+		)
+		if err != nil {
+			// If there are more addrs available, immediately try
+			// those.
+			nextAddr, iteratorErr := tower.Addresses.Next()
```

Comment:
Q: are we sure the iterator isn't advanced/reset elsewhere?

Reply:
yeah it is. My thinking about this is that in most cases, the tower will have 1 address that just works and this will be seen by all threads using the tower iterator when they call `Peak()`. Then, since all the threads using these iterators are best-effort, they will eventually try again at some point if the Peak address does in fact fail. 

What I could do though is add a `Copy` method to the address iterator so that this method acts on a completely independent iterator. The only thing that that will then mean is that any addresses added while this method is running wont be seen. But again, since it is best effort and will try again on restart, that isnt such a big deal. 

Thoughts?

Reply:
added! 🚀